File: legacy/routes/dashboards.js
const express = require('express');
const router = express.Router();
const db = require('../database');
const {
  buildStatsWhereClause,
  buildPanelDataQuery,
  buildFilterGroupsWhereClause,
  buildMonthSeries,
} = require('../helpers/queryBuilders');

// ─── Helper: map a DB row to the Dashboard shape ───────────────────────────

function rowToDashboard(row) {
  return {
    id: row.id,
    name: row.name,
    isDefault: row.is_default,
    dateRangeStart: row.date_range_start,
    dateRangeEnd: row.date_range_end,
    panelCount: row.panel_count !== undefined ? parseInt(row.panel_count) : undefined,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

function rowToPanel(row) {
  return {
    id: row.id,
    dashboardId: row.dashboard_id,
    title: row.title,
    chartType: row.chart_type,
    seriesMode: row.series_mode,
    netOrientation: row.net_orientation || null,
    legendOptions: row.legend_options || null,
    filterGroups: row.filter_groups || [],
    panelOrder: row.panel_order,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

// ─── Dashboard CRUD ─────────────────────────────────────────────────────────

// GET /api/dashboards — list all
router.get('/dashboards', async (req, res) => {
  try {
    const result = await db.query(`
      SELECT d.*,
        (SELECT COUNT(*) FROM dashboard_panels p WHERE p.dashboard_id = d.id) AS panel_count
      FROM dashboards d
      ORDER BY d.created_at ASC
    `);
    res.json(result.rows.map(rowToDashboard));
  } catch (err) {
    console.error('Error listing dashboards:', err);
    res.status(500).json({ error: 'Failed to list dashboards' });
  }
});

// POST /api/dashboards — create
router.post('/dashboards', async (req, res) => {
  try {
    const { id, name, isDefault, dateRangeStart, dateRangeEnd } = req.body;
    if (isDefault) {
      await db.query('UPDATE dashboards SET is_default = FALSE, updated_at = NOW()');
    }
    const result = await db.query(
      `INSERT INTO dashboards (id, name, is_default, date_range_start, date_range_end)
       VALUES ($1, $2, $3, $4, $5)
       RETURNING *`,
      [id, name, !!isDefault, dateRangeStart, dateRangeEnd]
    );
    res.status(201).json(rowToDashboard(result.rows[0]));
  } catch (err) {
    console.error('Error creating dashboard:', err);
    res.status(500).json({ error: 'Failed to create dashboard' });
  }
});

// PATCH /api/dashboards/:id — update name, date range, or set as default
router.patch('/dashboards/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { name, isDefault, dateRangeStart, dateRangeEnd } = req.body;

    if (isDefault) {
      await db.query('UPDATE dashboards SET is_default = FALSE, updated_at = NOW()');
    }

    const fields = [];
    const params = [];
    let idx = 1;

    if (name !== undefined) {
      fields.push(`name = $${idx++}`);
      params.push(name);
    }
    if (isDefault !== undefined) {
      fields.push(`is_default = $${idx++}`);
      params.push(!!isDefault);
    }
    if (dateRangeStart !== undefined) {
      fields.push(`date_range_start = $${idx++}`);
      params.push(dateRangeStart);
    }
    if (dateRangeEnd !== undefined) {
      fields.push(`date_range_end = $${idx++}`);
      params.push(dateRangeEnd);
    }
    fields.push(`updated_at = NOW()`);

    if (fields.length === 1) return res.status(400).json({ error: 'Nothing to update' });

    params.push(id);
    const result = await db.query(
      `UPDATE dashboards SET ${fields.join(', ')} WHERE id = $${idx} RETURNING *`,
      params
    );
    if (result.rows.length === 0) return res.status(404).json({ error: 'Dashboard not found' });
    res.json(rowToDashboard(result.rows[0]));
  } catch (err) {
    console.error('Error updating dashboard:', err);
    res.status(500).json({ error: 'Failed to update dashboard' });
  }
});

// DELETE /api/dashboards/:id
router.delete('/dashboards/:id', async (req, res) => {
  try {
    await db.query('DELETE FROM dashboards WHERE id = $1', [req.params.id]);
    res.json({ success: true });
  } catch (err) {
    console.error('Error deleting dashboard:', err);
    res.status(500).json({ error: 'Failed to delete dashboard' });
  }
});

// ─── Panel CRUD ──────────────────────────────────────────────────────────────

// GET /api/dashboards/:id/panels — list panels
router.get('/dashboards/:id/panels', async (req, res) => {
  try {
    const result = await db.query(
      'SELECT * FROM dashboard_panels WHERE dashboard_id = $1 ORDER BY panel_order ASC',
      [req.params.id]
    );
    res.json(result.rows.map(rowToPanel));
  } catch (err) {
    console.error('Error listing panels:', err);
    res.status(500).json({ error: 'Failed to list panels' });
  }
});

// POST /api/dashboards/:id/panels — create panel
router.post('/dashboards/:id/panels', async (req, res) => {
  try {
    const { id: dashboardId } = req.params;
    const {
      id,
      title,
      chartType,
      filterGroups,
      seriesMode,
      netOrientation,
      legendOptions,
      panelOrder,
    } = req.body;

    // Enforce 15-panel limit
    const countResult = await db.query(
      'SELECT COUNT(*) FROM dashboard_panels WHERE dashboard_id = $1',
      [dashboardId]
    );
    if (parseInt(countResult.rows[0].count) >= 15) {
      return res.status(400).json({ error: 'Dashboard has reached the 15-panel limit' });
    }

    const result = await db.query(
      `INSERT INTO dashboard_panels
         (id, dashboard_id, title, chart_type, filter_groups, series_mode, net_orientation, legend_options, panel_order)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
       RETURNING *`,
      [
        id,
        dashboardId,
        title,
        chartType,
        JSON.stringify(filterGroups || []),
        seriesMode || 'two_series',
        netOrientation || null,
        legendOptions ? JSON.stringify(legendOptions) : null,
        panelOrder || 0,
      ]
    );
    res.status(201).json(rowToPanel(result.rows[0]));
  } catch (err) {
    console.error('Error creating panel:', err);
    res.status(500).json({ error: 'Failed to create panel' });
  }
});

// POST /api/dashboard-panels/preview — MUST be before /:panelId
router.post('/dashboard-panels/preview', async (req, res) => {
  try {
    const { filterGroups, userId, dateFrom, dateTo, limit = 10, offset = 0 } = req.body;

    const { whereSql: baseSql, params } = buildStatsWhereClause(dateFrom, dateTo, userId);
    let nextParam = params.length + 1;

    const { sql: filterSql, nextParam: updatedParam } = buildFilterGroupsWhereClause(
      filterGroups || [],
      params,
      nextParam
    );
    nextParam = updatedParam;

    const extraConditions = filterSql ? ` AND ${filterSql}` : '';
    const fullWhere = `${baseSql}${extraConditions}`;

    const countResult = await db.query(`SELECT COUNT(*) FROM transactions ${fullWhere}`, params);
    const total = parseInt(countResult.rows[0].count);

    const limitParam = nextParam;
    const offsetParam = nextParam + 1;
    const dataResult = await db.query(
      `SELECT id, date, description, category, amount, type, user_id
       FROM transactions
       ${fullWhere}
       ORDER BY date DESC
       LIMIT $${limitParam} OFFSET $${offsetParam}`,
      [...params, parseInt(limit), parseInt(offset)]
    );

    const transactions = dataResult.rows.map((row) => ({
      id: row.id,
      date: row.date,
      description: row.description,
      category: row.category,
      amount: parseFloat(row.amount),
      type: row.type,
      user: row.user_id,
    }));

    res.json({ transactions, total });
  } catch (err) {
    console.error('Error previewing panel transactions:', err);
    res.status(500).json({ error: 'Failed to preview transactions' });
  }
});

// POST /api/dashboard-panels/chart-preview — monthly aggregates for filter preview
router.post('/dashboard-panels/chart-preview', async (req, res) => {
  try {
    const { filterGroups, userId, dateFrom, dateTo } = req.body;

    const { sql, params } = buildPanelDataQuery({
      dateFrom,
      dateTo,
      userId: userId || null,
      filterGroups: filterGroups || [],
    });

    const result = await db.query(sql, params);

    const monthMap = buildMonthSeries(dateFrom, dateTo);
    const rows = result.rows.map((row) => ({
      sortMonth: row.sort_month,
      month: row.month,
      type: row.type,
      total: parseFloat(row.total),
    }));

    res.json({ rows, monthMap });
  } catch (err) {
    console.error('Error generating chart preview:', err);
    res.status(500).json({ error: 'Failed to generate chart preview' });
  }
});

// PATCH /api/dashboard-panels/:panelId — update panel config
router.patch('/dashboard-panels/:panelId', async (req, res) => {
  try {
    const { panelId } = req.params;
    const {
      title,
      chartType,
      filterGroups,
      seriesMode,
      netOrientation,
      legendOptions,
      panelOrder,
    } = req.body;

    const fields = [];
    const params = [];
    let idx = 1;

    if (title !== undefined) {
      fields.push(`title = $${idx++}`);
      params.push(title);
    }
    if (chartType !== undefined) {
      fields.push(`chart_type = $${idx++}`);
      params.push(chartType);
    }
    if (filterGroups !== undefined) {
      fields.push(`filter_groups = $${idx++}`);
      params.push(JSON.stringify(filterGroups));
    }
    if (seriesMode !== undefined) {
      fields.push(`series_mode = $${idx++}`);
      params.push(seriesMode);
    }
    if (netOrientation !== undefined) {
      fields.push(`net_orientation = $${idx++}`);
      params.push(netOrientation || null);
    }
    if (legendOptions !== undefined) {
      fields.push(`legend_options = $${idx++}`);
      params.push(legendOptions ? JSON.stringify(legendOptions) : null);
    }
    if (panelOrder !== undefined) {
      fields.push(`panel_order = $${idx++}`);
      params.push(panelOrder);
    }
    fields.push('updated_at = NOW()');

    if (fields.length === 1) return res.status(400).json({ error: 'Nothing to update' });

    params.push(panelId);
    const result = await db.query(
      `UPDATE dashboard_panels SET ${fields.join(', ')} WHERE id = $${idx} RETURNING *`,
      params
    );
    if (result.rows.length === 0) return res.status(404).json({ error: 'Panel not found' });
    res.json(rowToPanel(result.rows[0]));
  } catch (err) {
    console.error('Error updating panel:', err);
    res.status(500).json({ error: 'Failed to update panel' });
  }
});

// DELETE /api/dashboard-panels/:panelId
router.delete('/dashboard-panels/:panelId', async (req, res) => {
  try {
    await db.query('DELETE FROM dashboard_panels WHERE id = $1', [req.params.panelId]);
    res.json({ success: true });
  } catch (err) {
    console.error('Error deleting panel:', err);
    res.status(500).json({ error: 'Failed to delete panel' });
  }
});

// PATCH /api/dashboards/:id/panel-order — reorder all panels
router.patch('/dashboards/:id/panel-order', async (req, res) => {
  try {
    const { id: dashboardId } = req.params;
    const { panelIds } = req.body; // ordered array of panel IDs

    await Promise.all(
      panelIds.map((panelId, index) =>
        db.query(
          'UPDATE dashboard_panels SET panel_order = $1, updated_at = NOW() WHERE id = $2 AND dashboard_id = $3',
          [index, panelId, dashboardId]
        )
      )
    );

    res.json({ success: true });
  } catch (err) {
    console.error('Error reordering panels:', err);
    res.status(500).json({ error: 'Failed to reorder panels' });
  }
});

// ─── Data endpoint ───────────────────────────────────────────────────────────

// POST /api/dashboards/:id/data — batched monthly aggregate for all panels
router.post('/dashboards/:id/data', async (req, res) => {
  try {
    const { id: dashboardId } = req.params;
    const { userId, dateRangeStart, dateRangeEnd } = req.body;

    // Load all panels for this dashboard
    const panelsResult = await db.query(
      'SELECT * FROM dashboard_panels WHERE dashboard_id = $1 ORDER BY panel_order ASC',
      [dashboardId]
    );
    const panels = panelsResult.rows.map(rowToPanel);

    // Run one aggregate query per panel in parallel
    const panelDataResults = await Promise.all(
      panels.map(async (panel) => {
        const { sql, params } = buildPanelDataQuery({
          dateFrom: dateRangeStart,
          dateTo: dateRangeEnd,
          userId: userId || null,
          filterGroups: panel.filterGroups,
        });

        const result = await db.query(sql, params);

        const monthMap = buildMonthSeries(dateRangeStart, dateRangeEnd);

        // Aggregate rows into per-month data
        for (const row of result.rows) {
          const key = row.sort_month;
          if (!monthMap[key]) monthMap[key] = { month: row.month };
          const total = parseFloat(row.total);
          if (panel.seriesMode === 'net_amount') {
            const sign = row.type === 'income' ? 1 : -1;
            monthMap[key].net = (monthMap[key].net || 0) + sign * total;
          } else {
            if (row.type === 'income') monthMap[key].income = total;
            else monthMap[key].expenses = total;
          }
        }

        return {
          panelId: panel.id,
          data: Object.entries(monthMap)
            .sort(([a], [b]) => a.localeCompare(b))
            .map(([, v]) => v),
        };
      })
    );

    res.json({ panels: panelDataResults });
  } catch (err) {
    console.error('Error fetching dashboard data:', err);
    res.status(500).json({ error: 'Failed to fetch dashboard data' });
  }
});

module.exports = router;

File: backend/app/routes/dashboards.py
from datetime import UTC, datetime
from datetime import date as date_type

from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy import delete, func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.dashboard import Dashboard, DashboardPanel
from app.models.transaction import Transaction
from app.schemas.dashboard import (
    ChartPreviewRequest,
    DashboardCreateRequest,
    DashboardDataRequest,
    DashboardOut,
    DashboardUpdateRequest,
    PanelCreateRequest,
    PanelOrderRequest,
    PanelOut,
    PanelPreviewRequest,
    PanelUpdateRequest,
)
from app.utils.query_builder import (
    build_filter_groups_clause,
    build_month_series,
    build_panel_data_query,
    build_stats_filter,
)

router = APIRouter(prefix="/api", tags=["dashboards"])


@router.get("/dashboards")
async def list_dashboards(db: AsyncSession = Depends(get_db)):
    panel_count_subq = (
        select(func.count())
        .where(DashboardPanel.dashboard_id == Dashboard.id)
        .correlate(Dashboard)
        .scalar_subquery()
    )
    stmt = select(Dashboard, panel_count_subq.label("panel_count")).order_by(
        Dashboard.created_at.asc()
    )
    result = await db.execute(stmt)
    rows = result.all()
    return [DashboardOut.from_orm_model(row[0], panel_count=row[1]) for row in rows]


@router.post("/dashboards", status_code=201)
async def create_dashboard(
    body: DashboardCreateRequest,
    db: AsyncSession = Depends(get_db),
):
    if body.isDefault:
        await db.execute(
            update(Dashboard).values(
                is_default=False,
                updated_at=datetime.now(UTC).replace(tzinfo=None),
            )
        )

    dashboard = Dashboard(
        id=body.id,
        name=body.name,
        is_default=body.isDefault,
        date_range_start=date_type.fromisoformat(body.dateRangeStart),
        date_range_end=date_type.fromisoformat(body.dateRangeEnd),
    )
    db.add(dashboard)
    await db.commit()
    await db.refresh(dashboard)
    return DashboardOut.from_orm_model(dashboard)


@router.patch("/dashboards/{dashboard_id}")
async def update_dashboard(
    dashboard_id: str,
    body: DashboardUpdateRequest,
    db: AsyncSession = Depends(get_db),
):
    update_data = body.model_dump(exclude_unset=True)
    if not update_data:
        return JSONResponse(
            status_code=400,
            content={"error": "Nothing to update"},
        )

    result = await db.execute(select(Dashboard).where(Dashboard.id == dashboard_id))
    dashboard = result.scalar_one_or_none()
    if not dashboard:
        return JSONResponse(
            status_code=404,
            content={"error": "Dashboard not found"},
        )

    if body.isDefault:
        await db.execute(
            update(Dashboard).values(
                is_default=False,
                updated_at=datetime.now(UTC).replace(tzinfo=None),
            )
        )

    if body.name is not None:
        dashboard.name = body.name
    if body.isDefault is not None:
        dashboard.is_default = body.isDefault
    if body.dateRangeStart is not None:
        dashboard.date_range_start = date_type.fromisoformat(body.dateRangeStart)
    if body.dateRangeEnd is not None:
        dashboard.date_range_end = date_type.fromisoformat(body.dateRangeEnd)
    dashboard.updated_at = datetime.now(UTC).replace(tzinfo=None)

    await db.commit()
    await db.refresh(dashboard)
    return DashboardOut.from_orm_model(dashboard)


@router.delete("/dashboards/{dashboard_id}")
async def delete_dashboard(
    dashboard_id: str,
    db: AsyncSession = Depends(get_db),
):
    await db.execute(delete(Dashboard).where(Dashboard.id == dashboard_id))
    await db.commit()
    return {"success": True}


@router.get("/dashboards/{dashboard_id}/panels")
async def list_panels(
    dashboard_id: str,
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(DashboardPanel)
        .where(DashboardPanel.dashboard_id == dashboard_id)
        .order_by(DashboardPanel.panel_order.asc())
    )
    return [PanelOut.from_orm_model(p) for p in result.scalars().all()]


@router.post("/dashboards/{dashboard_id}/panels", status_code=201)
async def create_panel(
    dashboard_id: str,
    body: PanelCreateRequest,
    db: AsyncSession = Depends(get_db),
):
    count_result = await db.execute(
        select(func.count()).where(DashboardPanel.dashboard_id == dashboard_id)
    )
    if count_result.scalar() >= 15:
        return JSONResponse(
            status_code=400,
            content={"error": "Dashboard has reached the 15-panel limit"},
        )

    panel = DashboardPanel(
        id=body.id,
        dashboard_id=dashboard_id,
        title=body.title,
        chart_type=body.chartType,
        filter_groups=body.filterGroups,
        series_mode=body.seriesMode,
        net_orientation=body.netOrientation,
        legend_options=body.legendOptions,
        panel_order=body.panelOrder,
    )
    db.add(panel)
    await db.commit()
    await db.refresh(panel)
    return PanelOut.from_orm_model(panel)


@router.patch("/dashboard-panels/{panel_id}")
async def update_panel(
    panel_id: str,
    body: PanelUpdateRequest,
    db: AsyncSession = Depends(get_db),
):
    update_data = body.model_dump(exclude_unset=True)
    if not update_data:
        return JSONResponse(
            status_code=400,
            content={"error": "Nothing to update"},
        )

    result = await db.execute(
        select(DashboardPanel).where(DashboardPanel.id == panel_id)
    )
    panel = result.scalar_one_or_none()
    if not panel:
        return JSONResponse(
            status_code=404,
            content={"error": "Panel not found"},
        )

    if body.title is not None:
        panel.title = body.title
    if body.chartType is not None:
        panel.chart_type = body.chartType
    if body.filterGroups is not None:
        panel.filter_groups = body.filterGroups
    if body.seriesMode is not None:
        panel.series_mode = body.seriesMode
    if body.netOrientation is not None:
        panel.net_orientation = body.netOrientation
    if body.legendOptions is not None:
        panel.legend_options = body.legendOptions
    if body.panelOrder is not None:
        panel.panel_order = body.panelOrder
    panel.updated_at = datetime.now(UTC).replace(tzinfo=None)

    await db.commit()
    await db.refresh(panel)
    return PanelOut.from_orm_model(panel)


@router.delete("/dashboard-panels/{panel_id}")
async def delete_panel(
    panel_id: str,
    db: AsyncSession = Depends(get_db),
):
    await db.execute(delete(DashboardPanel).where(DashboardPanel.id == panel_id))
    await db.commit()
    return {"success": True}


@router.patch("/dashboards/{dashboard_id}/panel-order")
async def reorder_panels(
    dashboard_id: str,
    body: PanelOrderRequest,
    db: AsyncSession = Depends(get_db),
):
    for index, panel_id in enumerate(body.panelIds):
        await db.execute(
            update(DashboardPanel)
            .where(
                DashboardPanel.id == panel_id,
                DashboardPanel.dashboard_id == dashboard_id,
            )
            .values(
                panel_order=index,
                updated_at=datetime.now(UTC).replace(tzinfo=None),
            )
        )
    await db.commit()
    return {"success": True}


@router.post("/dashboard-panels/preview")
async def panel_preview(
    body: PanelPreviewRequest,
    db: AsyncSession = Depends(get_db),
):
    filters = build_stats_filter(body.dateFrom, body.dateTo, body.userId)

    fg_clause = build_filter_groups_clause(body.filterGroups)
    if fg_clause is not None:
        filters.append(fg_clause)

    count_stmt = select(func.count()).select_from(Transaction).where(*filters)
    count_result = await db.execute(count_stmt)
    total = count_result.scalar()

    data_stmt = (
        select(
            Transaction.id,
            Transaction.date,
            Transaction.description,
            Transaction.category,
            Transaction.amount,
            Transaction.type,
            Transaction.created_by_user_id,
        )
        .where(*filters)
        .order_by(Transaction.date.desc())
        .limit(body.limit)
        .offset(body.offset)
    )
    data_result = await db.execute(data_stmt)
    transactions = [
        {
            "id": row.id,
            "date": str(row.date),
            "description": row.description,
            "category": row.category,
            "amount": float(row.amount),
            "type": row.type,
            "user": row.created_by_user_id,
        }
        for row in data_result.all()
    ]

    return {"transactions": transactions, "total": total}


@router.post("/dashboard-panels/chart-preview")
async def chart_preview(
    body: ChartPreviewRequest,
    db: AsyncSession = Depends(get_db),
):
    stmt, _ = build_panel_data_query(
        date_from=body.dateFrom,
        date_to=body.dateTo,
        user_id=body.userId,
        filter_groups=body.filterGroups,
    )

    result = await db.execute(stmt)
    month_map = build_month_series(body.dateFrom or "", body.dateTo or "")

    rows = [
        {
            "sortMonth": row.sort_month,
            "month": row.month,
            "type": row.type,
            "total": float(row.total),
        }
        for row in result.all()
    ]

    return {"rows": rows, "monthMap": month_map}


@router.post("/dashboards/{dashboard_id}/data")
async def dashboard_data(
    dashboard_id: str,
    body: DashboardDataRequest,
    db: AsyncSession = Depends(get_db),
):
    panel_result = await db.execute(
        select(DashboardPanel)
        .where(DashboardPanel.dashboard_id == dashboard_id)
        .order_by(DashboardPanel.panel_order.asc())
    )
    panels = [PanelOut.from_orm_model(p) for p in panel_result.scalars().all()]

    panel_data_results = []
    for panel in panels:
        stmt, _ = build_panel_data_query(
            date_from=body.dateRangeStart,
            date_to=body.dateRangeEnd,
            user_id=body.userId,
            filter_groups=panel.filterGroups,
        )

        result = await db.execute(stmt)
        month_map = build_month_series(
            body.dateRangeStart or "", body.dateRangeEnd or ""
        )

        for row in result.all():
            key = row.sort_month
            if key not in month_map:
                month_map[key] = {"month": row.month}
            total = float(row.total)
            if panel.seriesMode == "net_amount":
                sign = 1 if row.type == "income" else -1
                month_map[key]["net"] = month_map[key].get("net", 0) + sign * total
            else:
                if row.type == "income":
                    month_map[key]["income"] = total
                else:
                    month_map[key]["expenses"] = total

        sorted_data = [v for _, v in sorted(month_map.items(), key=lambda x: x[0])]

        panel_data_results.append(
            {
                "panelId": panel.id,
                "data": sorted_data,
            }
        )

    return {"panels": panel_data_results}
